ngraph/core/src/op/proposal.cpp

```diff
@@ -37,51 +37,85 @@ op::v0::Proposal::Proposal(const Output<Node>& class_probs,
+
+    NODE_VALIDATION_CHECK(this,
+                          image_shape_ps.rank().compatible(1),
+                          "Proposal layer shape image_shape should be rank 1 compatible (",
+                          image_shape_ps,
+                          ").");
+
+    if (bbox_deltas_ps.is_static() && class_probs_ps.is_static())
     {
-        const Shape class_probs_shape{class_probs_pshape.to_shape()};
-        const Shape class_bbox_deltas_shape{class_bbox_deltas_pshape.to_shape()};
-        const Shape image_shape_shape{image_shape_pshape.to_shape()};
-
-        NODE_VALIDATION_CHECK(
-            this,
-            class_probs_shape.size() == 4,
-            "Proposal layer shape class_probs input must have rank 4 (class_probs_shape: ",
-            class_probs_shape,
-            ").");
-
+        // class probs and bbox deltas shapes are static, check anchor count and batch number
+        // consistency
         NODE_VALIDATION_CHECK(this,
-                              class_bbox_deltas_shape.size() == 4,
-                              "Proposal layer shape class_bbox_deltas_shape input must have rank 4 "
-                              "(class_bbox_deltas_shape: ",
-                              class_bbox_deltas_shape,
+                              class_probs_ps[1].get_length() * 2 == bbox_deltas_ps[1].get_length(),
+                              "Anchor number inconsistent between class_probs (",
+                              class_probs_ps[1].get_length() / 2,
+                              "), and bbox_deltas (",
+                              bbox_deltas_ps[1].get_length() / 4,
                               ").");
 
-        NODE_VALIDATION_CHECK(
-            this,
-            image_shape_shape.size() == 1,
-            "Proposal layer image_shape input must have rank 1 (image_shape_shape: ",
-            image_shape_shape,
-            ").");
+        NODE_VALIDATION_CHECK(this,
+                              class_probs_ps[0] == bbox_deltas_ps[0],
+                              "Batch size inconsistent between class_probs (",
+                              class_probs_ps[0],
+                              ") and bbox deltas (",
+                              bbox_deltas_ps[0],
+                              ").");
+    }
 
+    if (image_shape_ps.is_static())
+    {
         NODE_VALIDATION_CHECK(
             this,
-            image_shape_shape[0] >= 3 && image_shape_shape[0] <= 4,
+            image_shape_ps[0].get_length() >= 3 && image_shape_ps[0].get_length() <= 4,
             "Image_shape 1D tensor must have => 3 and <= 4 elements (image_shape_shape[0]",
-            image_shape_shape[0],
+            image_shape_ps[0],
             ").");
-
-        auto batch_size = class_probs_shape[0];
-        set_output_type(0, get_input_element_type(0), Shape{batch_size * m_attrs.post_nms_topn, 5});
     }
-    else
+
+    // intersect the batch size
+    Dimension batch_dim = class_probs_ps[0] & bbox_deltas_ps[0];
```

Comment:
Before using a particular dimension we should check that the particular partial shape rank is not dynamic. Please, add new unit tests to cover this scenario

Reply:
Even if both ranks are dynamic, shouldn't it result in a dynamic dimension output?